elf_load.c

```diff
@@ -311,13 +312,132 @@ static int elf_load_imgcpy(struct elf_prog *elf_prog, Elf *elf,
 }
 
 #if CONFIG_LIBVFSCORE
+#if CONFIG_LIBPOSIX_MMAP
+static int elf_load_fdphdr(struct elf_prog *elf_prog, GElf_Phdr *phdr, int fd)
+{
+	uintptr_t vastart, vaend;
+	uint64_t delta_p_offset;
+	static void *addr;
+	static int flags;
+
+	/* If we are mmapping a fresh ELF, restore address and flags */
+	if (!phdr->p_paddr) {
+		addr = NULL;
+		flags = MAP_PRIVATE;
+	}
+
+	/* PT_LOAD's paddr may be misaligned, so keep in mind the
+	 * offset from what should have been the aligned paddr to
+	 * what we ended up with.
+	 */
+	delta_p_offset = phdr->p_paddr - PAGE_ALIGN_DOWN(phdr->p_paddr);
+
+	/* If this is not the first mmap'd ELF segment, then keep
+	 * contiguously mapping from where we left off.
+	 */
+	if (elf_prog->vabase)
+		addr = (void *)PAGE_ALIGN_DOWN((phdr->p_paddr +
+					       (uintptr_t)elf_prog->vabase));
+
+	uk_pr_debug("%s: Memory mapping 0x%"PRIx64" - 0x%"PRIx64" to 0x%"PRIx64" - 0x%"PRIx64"\n",
+		    elf_prog->name,
+		    (uint64_t)phdr->p_offset,
+		    (uint64_t)phdr->p_filesz + delta_p_offset,
+		    (uint64_t)addr,
+		    (uint64_t)addr + (uint64_t)phdr->p_filesz +
+		    delta_p_offset);
+
+	/* mmap with all flags. If protections are enabled, these will
+	 * be manually re-adjusted later.
+	 */
+	vastart = (uintptr_t)mmap(addr, phdr->p_filesz + delta_p_offset,
+				  PROT_EXEC | PROT_READ | PROT_WRITE,
+				  flags,
```

Comment:
Also, there is not a `MAP_FIXED`. If the virtual address is occupied by something else, you will not be able to resolve the conflict, right?

Reply:
So here needs a `MAP_FIXED` and should exit for failure, if `mmap` fails?

Reply:
`flags` has the `MAP_FIXED` bit set a few lines below :thinking: is that not enough?

Reply:
Sorry, I just missed that. It's enough.